core/src/apps/cardano/helpers/lazy_cbor_collection.py

```diff
@@ -0,0 +1,130 @@
+    """
+
+    class PauseIteration:
+        """
+        Used to signal to consumers of LazyCborCollection that all available data has been consumed for now.
+        """
+
+        pass
+
+    class EmptyElement:
+        """
+        Serves as a replacement for None since None can be a valid CBOR item.
+        """
+
+        pass
+
+    def __init__(self, size: int) -> None:
+        self.size = size
+        self.added_elements_count = 0
+        self.element: T | LazyCborCollection.EmptyElement = self.EmptyElement()
+
+    def __len__(self) -> int:
+        return self.size
+
+    def is_filled(self) -> bool:
+        return self.added_elements_count == self.size
+
+    def __iter__(self) -> Iterator[T]:
+        return self
+
+    def __next__(self) -> T:
+        if self.is_filled() and isinstance(self.element, self.EmptyElement):
+            raise StopIteration
+        assert not isinstance(
+            self.element, self.EmptyElement
+        ), "The next element hasn't been added yet"
+        element, self.element = self.element, self.EmptyElement()
+        return element
+
+    def append_item(self, item: T) -> None:
+        assert isinstance(
+            self.element, self.EmptyElement
+        ), "The previous element hasn't been read yet"
+        assert (
+            self.added_elements_count < self.size
+        ), "Item appended to a filled collection"
+        self._validate_item(item)
+        self.element = item
+        self.added_elements_count += 1
+
+    def _validate_item(self, item: T) -> None:
+        pass
+
+    def cbor_serialize(self) -> Iterator[bytes | Type[PauseIteration]]:
+        """
+        Provides a custom encoder to lazy collections which has the highest priority when being encoded into CBOR.
+        This allows us to circumvent the default cbor serializer.
+        """
+        raise NotImplementedError
+
+
+class LazyCborDict(LazyCborCollection):
+    def _validate_item(self, item: tuple) -> None:
+        assert len(item) == 2
+
+    def cbor_serialize(
+        self,
+    ) -> Iterator[bytes | Type[LazyCborCollection.PauseIteration]]:
+        yield _get_cbor_header(_CBOR_TYPE_MAP, self.size)
+        yield self.PauseIteration
+
+        for element in self:
+            assert len(element) == 2
+            key, value = element
+            yield from cbor.encode_streamed(key)
+            yield from cbor.encode_streamed(value)
+            yield self.PauseIteration
+
+
+class LazyCborList(LazyCborCollection):
+    def cbor_serialize(
+        self,
+    ) -> Iterator[bytes | Type[LazyCborCollection.PauseIteration]]:
+        yield _get_cbor_header(_CBOR_TYPE_ARRAY, self.size)
+        yield self.PauseIteration
+
+        for element in self:
+            yield from cbor.encode_streamed(element)
+            yield self.PauseIteration
+
+
+def _get_cbor_header(typ: int, l: int) -> bytes:
```

Comment:
We weren't sure what the best approach here was. Either to make `cbor._header` public and thus making the cbor library less independent or to have a copy of the function.

Reply:
Rather than exposing the internal `_header` function I've created `create_array_header` and `create_map_header` functions which are "public". It seems more correct than exposing both the `_header` function and the type constants. Also the `_header` IMHO has a name which isn't descriptive enough for the outside world. I also added tests for these functions.

https://github.com/trezor/trezor-firmware/pull/1683/commits/45b53e8adebc8240d508e07df799b5e1a1de3ee8#diff-bce249b14d43a52eb8502d517376e3eebf19dcce888961d10dad2ac1218578caR315